logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/conversation/message/MLSWrongEpochHandler.kt

```diff
@@ -0,0 +1,100 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.sync.receiver.conversation.message
+
+import com.benasher44.uuid.uuid4
+import com.wire.kalium.logger.KaliumLogger
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.MLSFailure
+import com.wire.kalium.logic.data.conversation.Conversation
+import com.wire.kalium.logic.data.conversation.ConversationRepository
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageContent
+import com.wire.kalium.logic.data.message.PersistMessageUseCase
+import com.wire.kalium.logic.data.user.UserId
+import com.wire.kalium.logic.feature.conversation.JoinExistingMLSConversationUseCase
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.logic.kaliumLogger
+
+interface MLSWrongEpochHandler {
+    suspend fun onMLSWrongEpoch(
+        conversationId: ConversationId,
+        dateIso: String,
+    )
+}
+
+internal class MLSWrongEpochHandlerImpl(
+    private val selfUserId: UserId,
+    private val persistMessage: PersistMessageUseCase,
+    private val conversationRepository: ConversationRepository,
+    private val joinExistingMLSConversation: JoinExistingMLSConversationUseCase
+) : MLSWrongEpochHandler {
+
+    private val logger by lazy { kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.EVENT_RECEIVER) }
+
+    override suspend fun onMLSWrongEpoch(
+        conversationId: ConversationId,
+        dateIso: String,
+    ) {
+        logger.i("Handling MLS WrongEpoch result")
+        conversationRepository.baseInfoById(conversationId).flatMap {
+            val protocol = it.protocol
+            if (protocol is Conversation.ProtocolInfo.MLS) {
+                Either.Right(protocol)
+            } else {
+                Either.Left(MLSFailure.ConversationDoesNotSupportMLS)
+            }
+        }.flatMap { currentProtocol ->
+            getUpdatedConversationEpoch(conversationId).map { updatedEpoch ->
+                updatedEpoch != null && updatedEpoch != currentProtocol.epoch
+            }
+        }.flatMap { isRejoinNeeded ->
+            if (isRejoinNeeded) {
+                joinExistingMLSConversation(conversationId)
+            } else Either.Right(Unit)
+        }.flatMap {
+            insertInfoMessage(conversationId, dateIso)
+        }
+    }
+
+    private suspend fun getUpdatedConversationEpoch(conversationId: ConversationId): Either<CoreFailure, ULong?> {
+        return conversationRepository.fetchConversation(conversationId).flatMap {
```

Comment:
can we use smaller query and data flow here to fetch what we need?

Reply:
I didn't notice we already had a function for it. I'll change it to use it.

Otherwise, considering that it's fetching a single conversation (and not a list), I wouldn't go through the trouble to create a new query, add tests, etc. as the performance improvement is pretty much non-existent.